metrics/metrics-tests/src/main/java/org/apache/james/metrics/tests/RecordingMetricFactory.java

```diff
@@ -0,0 +1,79 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.metrics.tests;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+import org.apache.james.metrics.api.TimeMetric;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+public class RecordingMetricFactory implements MetricFactory {
+    private final Multimap<String, Duration> executionTimesInMs = Multimaps.synchronizedSetMultimap(HashMultimap.create());
+    private final ConcurrentHashMap<String, Integer> executionValues = new ConcurrentHashMap<>();
+
+    @Override
+    public Metric generate(String name) {
+        return new RecordingMetric(name, executionValue -> executionValues.put(name, executionValue));
+    }
+
+    @Override
+    public TimeMetric timer(String name) {
+        return new RecordingTimeMetric(name, executionTime -> {
+            synchronized (executionTimesInMs) {
+                executionTimesInMs.put(name, executionTime);
+            }
+        });
+    }
+
+    public Collection<Duration> executionTimesFor(String name) {
+        synchronized (executionTimesInMs) {
+            return executionTimesInMs.get(name);
+        }
+    }
+
+    public Multimap<String, Collection<Duration>> executionTimesForPrefixName(String prefixName) {
+        synchronized (executionTimesInMs) {
+            return executionTimesInMs.asMap().entrySet().stream()
+                .filter(entry -> entry.getKey().startsWith(prefixName))
+                .collect(Multimaps.toMultimap(Map.Entry::getKey, Map.Entry::getValue, HashMultimap::create));
+        }
+    }
+
+    public int countFor(String name) {
+        return executionValues.getOrDefault(name, 0);
+    }
+
+    public List<Integer> countForPrefixName(String prefixName) {
```

Comment:
returning an Integer is better IMHO, because it's a count :)

Reply:
but if we have more metrics with same prefixName?

Reply:
Then maybe a map of matching entries?